### ISSUE TITLE
Update all deriving statements to use DerivingVia style

### ISSUE THREAD

**Issue by recursion-ninja** [amnh/PCG#148]
The [`DerivingVia`][0] language extension allows for a much more descriptive `deriving` statement, and in some cases, can remove the need for explicit `instance` declarations.

The following example shows how by using the `DerivingVia` style of `deriving` statement, it is clear *how* the instance below is derived from the GHC stock deriving mechanism:

    data Foo = A | B | C
      deriving stock (Eq, Ord, Show)

The following example shows similarly how we can specify deriving instances by coercing a newtype:

    newtype Bar f a = Bar (f a) 
      deriving newtype (Functor, Applicative)

We should replace all the existing `deriving` statements in our codebase with the new `DerivingVia` style for improved clarity of the deriving mechanism used.

Additionally, we should inspect existing instance declarations for opportunities to remove tedious wrapping & unwrapping and/or coercion of `newtype`s and replace the the instance with a deriving via mechanism.

[0]: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-DerivingVia

**Comment by Boarders** [comment]
This is conflating some different issues. The examples of the form:
```haskell
data Foo = A | B | C
  deriving stock (Eq, Ord, Show)
```
are from the [`DerivingStrategies`](https://gitlab.haskell.org/ghc/ghc/wikis/commentary/compiler/deriving-strategies) language extension. In newer versions of hlint (after hlint-2-1-25) this can be enabled as a warning and I have done so [here](https://github.com/amnh/PCG/commit/f720caf2e67b45f742a971dd7a567fdcbf21908b). This commit also adds the deriving strategy we have used to all newtype declarations in our code base and adds a few easy-to-spot instances that can be derived rather than written out explicitly. `DerivingVia` goes one step further and allows one to do something like the following:
```haskell
newtype Index = Index {getIndex :: Int}
  deriving Monoid (via Sum Int)
```
I can imagine this being quite useful in our code base but it is much less mechanical to figure out where it can be employed.

**Comment by recursion-ninja** [comment]
Yes, this is what I intended. I thought `DerivingVia` & `DerivingStrategies` were a single language pragma. 

What you stated is what I intended.

At some point (it's low priority) we should look for the less mechanical instances that could be replaced with `deriving via`.

**Comment by recursion-ninja** [comment]
We can add the new [`-Wmissing-deriving-strategies`](https://downloads.haskell.org/ghc/8.8.1-rc1/docs/html/users_guide/using-warnings.html#ghc-flag--Wmissing-deriving-strategies) flag once our code compiler with GHC 8.8.1.
